std/algorithm/iteration.d

```diff
@@ -4289,3 +4299,484 @@ unittest
+
+
+size_t binomial(size_t n, size_t k) pure nothrow @safe @nogc
+in
+{
+    assert(n > 0, "binomial: n must be > 0.");
+}
+body
+{
+    if (k < 0 || k > n)
+        return 0;
+    if (k == n)
+        return 1;
+    if (k > (n / 2))
+        k = n - k;
+    size_t result = 1;
+    for (size_t i=1, m=n; i<=k; i++, m--)
+        result = result * m/i;
+    return result;
+}
+
+///
+pure nothrow @safe @nogc unittest
+{
+    assert(binomial(5,2) == 10);
+    assert(binomial(6,4) == 15);
+    assert(binomial(3,1) == 3);
+}
+
+pure nothrow @safe @nogc unittest
+{
+    assert(binomial(5,1) == 5);
+    assert(binomial(5,0) == 1);
+    assert(binomial(1,2) == 0);
+    assert(binomial(1,0) == 1);
+    assert(binomial(1,1) == 1);
+    assert(binomial(2,1) == 2);
+}
+
+/**
+Lazily computes the Cartesian power of $(D r)
+for a number of repetitions $(D repeat).
+If the input is sorted, the product is in lexicographic order.
+For example $(D"AB".cartesianPower(2).array) returns $(D["AA", "AB", "BA", "BB"])
+
+Params:
+    r = RandomAccessRange or string as origin
+    repeat = number of repetitions
+
+Returns:
+    Forward range which yields the product items
+*/
+auto cartesianPower(Range)(Range r, size_t repeat=1)
+if (isRandomAccessRange!Range && hasLength!Range)
+in {
+    assert(repeat >= 1, "Invalid repeat");
+}
+body
+{
+    import std.array: array;
+    import std.range: indexed;
+    static struct CartesianPower
+    {
+        ElementType!Range[] front;
+        private Range _r;
+        private size_t[] _state;
+
+        private size_t _repeat, _nrElements;
+        private size_t _max_states, _pos;
+        private bool _isNrStatesComputed;
+
+        bool empty;
+
+        this(Range r, size_t repeat) pure nothrow @safe
+        {
+            _r = r;
+            // cache number of elements
+            _nrElements = r.length;
+            _repeat = repeat;
+
+            // set initial state and calculate max possibilities
+            _max_states = length;
+            if (_nrElements > 0 && _max_states > 0)
+            {
+                _state.length = repeat;
+                front = _r.indexed(_state).array;
+            }
+            else
+                empty = true;
+        }
+
+        @property typeof(this) save()
+        {
+            // we need to save more information about the current state
+            auto c = typeof(this)(_r, _repeat);
+            c._state = _state.dup;
+            c._pos = _pos;
+            c.empty = empty;
+            return c;
+        }
+
+        void popFront() pure nothrow @safe
+        {
+            if (empty) return;
+            // check whether we have reached the end
+            _pos++;
+            if (_pos == _max_states)
+            {
+                empty = true;
+                return;
+            }
+            foreach_reverse (i, ref el; _state)
+            {
+                ++el;
+                if (el < _nrElements)
+                    break;
+                else
```

Comment:
Because the behaviour of the code doesn't change if you remove it. But IMO that's ok here, because it is more explicit.